smack-extensions/src/main/java/org/jivesoftware/smackx/usertune/element/UserTuneElement.java

```diff
@@ -157,43 +160,53 @@ public UserTuneElement getUserTune() {
 
         private Map<Integer,String> ORDER = new HashMap<Integer,String>();
 
-        public void setArtist(String artist) {
+        private Builder() {
+        }
+
+        public Builder setArtist(String artist) {
             this.artist = artist;
             ORDER.put(++orderIndex,"artist");
+            return this;
         }
 
-        public void setLength(int length) {
+        public Builder setLength(int length) {
             this.length = length;
             ORDER.put(++orderIndex,"length");
+            return this;
         }
 
-        public void setRating(int rating) {
+        public Builder setRating(int rating) {
             if (rating <= 10 && rating >= 0) {
                 this.rating = rating;
                 ORDER.put(++orderIndex,"rating");
             } else {
                 throw new IllegalArgumentException("Rating tried to set was out of bounds");
             }
+            return this;
         }
 
-        public void setSource(String source) {
+        public Builder setSource(String source) {
             this.source = source;
             ORDER.put(++orderIndex,"source");
```

Comment:
what do you need `ORDER` for? It doesn't matter in which order your elements are in the XML.
You can simply define your own order of elements in toXML() by serializing "source" first, then artist, etc. (in any order you like). There shouldn't be a reason to preserve the order in which the user placed the values in the builder.

Reply:
Yes you are right that we do not need to preserve the order in which those elements are received or sent, but at the same time it is important to remember which elements are to be send and which are not.
Not storing them will result in an XML of the form
	  <tune xmlns='http://jabber.org/protocol/tune'>
	          <artist>Yes</artist>
	          <length></length>
	          <rating></rating>
	          <source>Yessongs</source>
	          <title>Heart of the Sunrise</title>
	          <track></track>
	   </tune>
Such a transmission will me meaningless.
Storing will result in
	<tune xmlns='http://jabber.org/protocol/tune'>
	          <artist>Yes</artist>
	          <source>Yessongs</source>
	          <title>Heart of the Sunrise</title>
	   </tune>

Reply:
Anyways there will surely be some way to do it without storing. Will figure it out and shortly.

Reply:
well, you can simply check
```
if (artist != null) {
    xml.apped("<artist>" + artist + "</artist>");
}
```
so you only add the opening and closing tags for attributes that are set.

Reply:
On the other hand, I'm not sure, if @Flowdalic will be pleased with manual opening xml tags via Strings like that. The proper solution would probably be to create dedicated `NamedElement`s, but on the other hand that might be overkill for this case...